### ISSUE TITLE
Multi-module project and ${project.version}

### ISSUE THREAD

**Issue by esiqveland** [jgitver/jgitver-maven-plugin#10]
I have a multi-module project, where inter-module dependencies are specified like this:

eg. in module `website-core` I depend on `website-domain`:

```
        <dependency>
            <groupId>com.business.website</groupId>
            <artifactId>website-domain</artifactId>
            <version>${project.version}</version>
        </dependency>
```

When building the parent of this, maven still sees this as version 0, and fails with error:

```
[ERROR] Failed to execute goal on project website-core: 
Could not resolve dependencies for project com.business.website:website-core:jar:2.5.0-1: 
The following artifacts could not be resolved: com.business.website:website-domain:jar:tests:0,
 com.business.website:website-integration:0: 
Failure to find com.business.website:website-domain:jar:tests:0 in 
http://nexus:8081/nexus/content/groups/public was cached in the local repository, resolution will not be 
reattempted until the update interval of nexus has elapsed or updates are forced -> [Help 1]
```

Should this work, or is my maven structure goofed?

**Comment by McFoggy** [comment]
it should work.

**Comment by esiqveland** [comment]
Ok, I will try to reproduce in an example.

**Comment by McFoggy** [comment]
@esiqveland currently there is at least one integration test for multi modules under `src/it/multi-pure-extension-it`, we could probably enhance with more

**Comment by esiqveland** [comment]
Ok, it works if i specify `${project.version}` in parent-pom for inter-module dependencies using `dependencyManagement`. The value of ${project.version} is still `0` in the child POM though.

Not sure if that is as intended.

**Comment by esiqveland** [comment]
It seems only dependencies with `<type>test-jar</type>` fail to resolve correct version.

**Comment by McFoggy** [comment]
Hmmm ; it starts to go deep in maven dependency resolution hell and it smells the maven bug.
Do you have a simple reproduce-able test case?

**Comment by esiqveland** [comment]
I have reproduced my issue using one of your example poms.
Check the jar produced by this build:

```
<project 
    xmlns="http://maven.apache.org/POM/4.0.0" 
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>fr.brouillard.oss.demo</groupId>
    <artifactId>demo-jgitver-maven-plugin</artifactId>
    <version>0</version>
    <packaging>jar</packaging>
    <build>
        <extensions>
            <extension>
                <groupId>fr.brouillard.oss</groupId>
                <artifactId>jgitver-maven-plugin</artifactId>
                <version>0.2.0-alpha2</version>
            </extension>
        </extensions>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.2</version>
                <configuration>
                    <source>1.8</source>
                    <target>1.8</target>
                    <encoding>UTF-8</encoding>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-jar-plugin</artifactId>
                <version>2.5</version>
                <configuration>
                    <finalName>${project.artifactId}-${project.version}-${jgitver.git_sha1_8}</finalName>
                    <archive>
                        <manifest>
                            <addDefaultImplementationEntries>true</addDefaultImplementationEntries>
                        </manifest>
                        <manifestEntries>
                            <App-Version>${project.artifactId}-${project.version}-${jgitver.git_sha1_8}</App-Version>
                        </manifestEntries>
                    </archive>
                </configuration>
            </plugin>
        </plugins>
    </build>
</project>
```

It will produce an artifact like so: `target/demo-jgitver-maven-plugin-0-26874b86.jar`

It seems like the jar gets installed using correct version:

```
[INFO] --- maven-install-plugin:2.4:install (default-install) @ demo-jgitver-maven-plugin ---
[INFO] Installing .../demo-jgitver-maven-plugin/target/demo-jgitver-maven-plugin-0-26874b86.jar to /Users/..../.m2/repository/fr/brouillard/oss/demo/demo-jgitver-maven-plugin/2.0.0-SNAPSHOT/demo-jgitver-maven-plugin-2.0.0-SNAPSHOT.jar
```

But the produced jar versioning is wonky.

```
mvn --version
Java HotSpot(TM) 64-Bit Server VM warning: ignoring option MaxPermSize=512m; support was removed in 8.0
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T17:41:47+01:00)
Maven home: /usr/local/Cellar/maven/3.3.9/libexec
Java version: 1.8.0_72, vendor: Oracle Corporation
Java home: /Library/Java/JavaVirtualMachines/jdk1.8.0_72.jdk/Contents/Home/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "mac os x", version: "10.11.2", arch: "x86_64", family: "mac"
```

**Comment by McFoggy** [comment]
For further investigation on maven property resolution 

```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
    <groupId>fr.brouillard.oss.demo</groupId>
    <artifactId>demo-jgitver-maven-plugin</artifactId>
    <version>0</version>
    <packaging>pom</packaging>
    <build>
        <extensions>
            <extension>
                <groupId>fr.brouillard.oss</groupId>
                <artifactId>jgitver-maven-plugin</artifactId>
                <version>0.2.0-alpha2</version>
            </extension>
        </extensions>
        <plugins>
            <plugin>
                <artifactId>maven-antrun-plugin</artifactId>
                <executions>
                    <execution>
                        <phase>validate</phase>
                        <goals>
                            <goal>run</goal>
                        </goals>
                        <configuration>
                            <tasks>
                                <echo>project version: ${project.version}</echo>
                            </tasks>
                        </configuration>
                    </execution>
                </executions>
            </plugin>        
        </plugins>
    </build>
</project>
```

## plugin help:evaluate: OK

`mvn help:evaluate -Dexpression=project.version` works as expected

```
...
[INFO] --- maven-help-plugin:2.2:evaluate (default-cli) @ demo-jgitver-maven-plugin ---
[INFO] No artifact parameter specified, using 'fr.brouillard.oss.demo:demo-jgitver-maven-plugin:pom:0.0.0-SNAPSHOT' as project.
[INFO]
0.0.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
...
```

## plugin ant: KO

```
$ mvn validate
...
[INFO] --- maven-antrun-plugin:1.3:run (default) @ demo-jgitver-maven-plugin ---
[INFO] Executing tasks
     [echo] project version: 0
[INFO] Executed tasks
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
...
```

**Comment by McFoggy** [comment]
using `maven-antrun-plugin:1.8` in above configuration and launching with `mvn -X validate` we see debug log from `AntRunMojo#copyProperties()` shows that version is correctly set:

```
[DEBUG] Setting properties with prefix:
Setting project property: project.groupId -> fr.brouillard.oss.demo
Setting project property: project.artifactId -> demo-jgitver-maven-plugin
Setting project property: project.name -> demo-jgitver-maven-plugin
Setting project property: project.version -> 0.0.0-SNAPSHOT
Setting project property: project.packaging -> pom
...
```

but still the evaluation of `${project.version}` in ant task is wrong

**Comment by mirkosertic** [comment]
My build is always breaking if using something like:

```
<dependencyManagement>
    <dependencies>
        <dependency>
            <groupId>...</groupId>
            <artifactId>...</artifactId>
            <version>${project.version}</version>
        </dependency>
    </dependencies>
</dependencyManagement>
```

for inter-project module dependencies inside of parent pom.xml

**Comment by McFoggy** [comment]
@mirkosertic this use case is under IT test at [jgitver/jgitver-maven-plugin/tree/master/src/it/multi-pure-extension-it](https://github.com/jgitver/jgitver-maven-plugin/tree/master/src/it/multi-pure-extension-it).
Can you please try to elaborate what is different in your setup?

**Comment by xeagle2** [comment]
The plugin is really useful, however If I build "multi-pure-extension-it" sample on Maven 3.3.9, the solution doesn't work as expected and the final file name contains 0 (my-module-0.jar) instead of proper version (my-module-1.0.0-SNAPSHOT.jar).

Plugin version: 0.2.0-alpha2

```
> mvn -v
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T17:41:47+01:00)
Maven home: c:\apps\maven\bin\..
Java version: 1.8.0_60, vendor: Oracle Corporation
Java home: c:\Program Files\Java\jdk1.8.0_60\jre
Default locale: en_US, platform encoding: Cp1251
OS name: "windows 10", version: "10.0", arch: "amd64", family: "dos"
```

**Comment by McFoggy** [comment]
@xeagle2 if you mean that you have `target\my-module-0.jar` instead of `target\my-module-X.Y.Z.jar` wher X.Y.Z is calculated by jgitver, I would say that in the current version it is normal.
Only `mvn install` & `mvn deploy` produce a good naming.  
This is because some part of the POM are calculated at the very beginning of the project build execution and that values are cached ; and thus resolved with 0 instead of X.Y.Z. Maven is an old software and not all the pieces & plugins use the same APIs to do the same thing leading in different behaviors from time to time.

**Comment by xeagle2** [comment]
@McFoggy, thanks for getting the clarified, agree with the Maven API. 

As far as ${project.version} is used somewhere and it is resolved to "0" it's not easy to rely on the filtering and variables, so to make it more clear in terms of behavior of the plugin I added one more property "jgitver.calculatedVersion" in forked version:

https://github.com/xeagle2/jgitver-maven-plugin/commit/5429750c4c6746024b98718c20e192a86330dbb6
https://github.com/xeagle2/jgitver-maven-plugin/commit/3773a0a6a4388b7d7544c3cd34ef9d5a59baaf8e

I believe it would be useful to have the similar in upstream version, wouldn't be?

**Comment by McFoggy** [comment]
@xeagle2 I agree it is the easiest to offer for now. Please open a PR & change to `jgitver.calculated_version` to be consistent with other properties.

In // we should continue to investigate how we could try to offer a better experience for jgtiver end users even if it means some more work to obtain a _consistent_ behavior (even if it is utopic) regarding  maven properties & filtering.

**Comment by danielwegener** [comment]
Just to mention it (kind of related). maven-assembly-plugin in a child module does also seem to pick up the empty `${project.version}` 0.

```
[INFO] --- maven-assembly-plugin:2.4:single (assembly) @ foo-bar ---
[INFO] Reading assembly descriptor: src/main/assembly.xml
[INFO] Building zip: /Users/daniel/projects/foo-bar/target/foo-bar-0.zip
```

**Comment by McFoggy** [comment]
What does your assembly.xml contain: `project.version` or `project.build.finalName`? If it is the latter then we are in the same situation than `mvn package` and indeed, in memory, `project.getBuild().getFinalName()` is resolved & cached very early in the build process.

**Comment by danielwegener** [comment]
IIRC it only had

```
<formats>
    <format>zip</format>
  </formats>
```

 (I'll have a look tomorrow)

**Comment by McFoggy** [comment]
this issue should now be corrected with the 0.3.0-alpha2 release.
Please review the README to use the new maven core extension mechnism.
